api/src/main/java/org/apache/iceberg/metrics/ScanReport.java

```diff
@@ -102,40 +113,304 @@ public Builder withSnapshotId(long newSnapshotId) {
       Preconditions.checkArgument(null != tableName, "Invalid table name: null");
-      Preconditions.checkArgument(null != filter, "Invalid expression filter: null");
       Preconditions.checkArgument(null != projection, "Invalid schema projection: null");
       Preconditions.checkArgument(null != scanMetrics, "Invalid scan metrics: null");
-      return new ScanReport(tableName, snapshotId, filter, projection, scanMetrics);
+      return new ScanReport(tableName, snapshotId, projection, scanMetrics);
+    }
+  }
+
+  /** A serializable version of a {@link Timer} that carries its result. */
+  public static class TimerResult implements Serializable {
+    private final String name;
+    private final TimeUnit timeUnit;
+    private final Duration totalDuration;
+    private final long count;
+
+    public TimerResult(String name, TimeUnit timeUnit, Duration totalDuration, long count) {
+      Preconditions.checkArgument(null != name, "Invalid timer name: null");
+      Preconditions.checkArgument(null != timeUnit, "Invalid time unit: null");
+      Preconditions.checkArgument(null != totalDuration, "Invalid duration: null");
+      this.name = name;
+      this.timeUnit = timeUnit;
+      this.totalDuration = totalDuration;
+      this.count = count;
+    }
+
+    public String name() {
+      return name;
+    }
+
+    public TimeUnit timeUnit() {
+      return timeUnit;
+    }
+
+    public Duration totalDuration() {
+      return totalDuration;
+    }
+
+    public long count() {
+      return count;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(name())
+          .add("duration", totalDuration())
+          .add("count", count)
+          .add("timeUnit", timeUnit)
+          .toString();
+    }
+
+    public static TimerResult fromTimer(Timer timer) {
+      Preconditions.checkArgument(null != timer, "Invalid timer: null");
+      return new TimerResult(timer.name(), timer.unit(), timer.totalDuration(), timer.count());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      TimerResult that = (TimerResult) o;
+      return count == that.count
+          && Objects.equal(name, that.name)
+          && timeUnit == that.timeUnit
+          && Objects.equal(totalDuration, that.totalDuration);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(name, timeUnit, totalDuration, count);
+    }
+  }
+
+  /**
+   * A serializable version of a {@link Counter} that carries its result.
+   *
+   * @param <T> The type of the {@link CounterResult}.
+   */
+  public static class CounterResult<T extends Number> implements Serializable {
+    private final String name;
+    private final MetricsContext.Unit unit;
+    private final T value;
+
+    public CounterResult(String name, Unit unit, T value) {
```

Comment:
I think we can make `CounterResult` always use a long. There should be no need for us to report metrics back as an Integer, especially if we are planning to remove the distinction from `Counter`.

Reply:
I wanted to do those changes after the Counter API gets refactored, but I guess we can make `CounterResult` always use a long now as well